designs/nodepool-termination-grace-period.md

```diff
@@ -0,0 +1,127 @@
+# NodePool Termination Grace Period
+
+## Motivation
+Users are requesting the ability to control how long karpenter will wait for the deletion of an individual node to complete before forcefully terminating the node regardless of pod status. [#743](https://github.com/kubernetes-sigs/karpenter/issues/743). This supports two primary use cases.
+* Cluster admins who want to ensure that nodes are cycled after a given period of time, regardless of user-defined disruption controls (such as PDBs or preStop hooks) that might prevent eviction of a pod beyond the configured limit. This could be to satisfy security requirements or for convenience.
```

Comment:
When you are forcefully draining nodes in this way, how do you coordinate with the app devs on your clusters? Do you inform them that this change is incoming? Is there a way that you can enforce that they don't do things like a fully-blocking PDB through policy or just discussing with them or is it just outright untenable to keep up with the number of ways that users might mis-configure their nodes?

Reply:
We have internal documentation dedicated to describing our node rollout processes for handling upgrades, how to control pod disruption with PDBs, preStop hooks, etc. We make it clear in these docs that voluntary disruptions must be tolerated by users of our clusters, and that they may delay eviction by up to X time before their pods are evicted forcefully (24h currently).

Reply:
> there needs to be some policy enforcement to ensure that users aren't setting this to an unnecessary high value or just to ensure that they aren't tolerating this taint at all and only let cluster admins selectively tolerate this taint.

1. I'm not sure if the out-of-service taint is handled specially by K8s. This can be tested.
2. If it's a problem, it can be solved by validating admission webhooks such as OPA and the like.

K8s has a lot of fields that are potentially disruptive to other users of a cluster, such as priorityClass, terminationGracePeriodSeconds, etc which often require additional validation controls to support multi-tenancy.